### PR TITLE
feat(mojaloop/#3472): enhance kafka lib (vNow) to inc perf stream recs

### PR DESCRIPTION
**Pull request by mdebarros**
!feat(mojaloop/#3472): enhance kafka lib (vNow) to inc perf stream recs - https://github.com/mojaloop/project/issues/3472
- added perf benchmarking tests for producer and consumer
- added new options 
    - `serializeFn` - default = `null` - override support for serialization (`serializeFn`) on Producer, with default using JSON.stringify.
    - `deserializeFn` - default = `null` - override support for de-serialzation (`deserializeFn`) on Consumer, with default using JSON.parse.
    - `syncConcurrency` - default = `1` (see below for more info)
    - `syncSingleMessage` - default = `false` (see below for more info)
- added new events:
    - `partition.eof` event - this is really helpful to determine once you have reached the end of a partition (i.e. no more messages) for Consumer only
    - `warning` event - this is really helpful to determine once you have reached the end of a partition (i.e. no more messages) for Consumer only
    - `event.stats` event - this will provide librdkafka stats based on the frequency configured by `statistics.interval.ms` (default = 0) rdkafkaConf value, with `0` disabling stats collection.
- added new features:
    - added `isConnected` method to both Producer and Consumers
    - added `connectedTime` method to both Producer and Consumers
    - added `version` method to both Producer and Consumers
    - added `getMetadaSync` method to both Producer and Consumers
- performance enhancements
     - removed internal on(`data`) event handler which was unnecessarily parsing the consumed Kafka message a 2nd time. Instead, each of the consumer implementation modes will emit('message') directly.
     - added support for `syncConcurrency` - this allows one to set the syncQueue processing concurrency
     - added support for `syncSingleMessage` - when set to true, "batch" messages will be queued up individually, and processed concurrently based on the `syncConcurrency`. When false, the "batch" of messages will be queued as-is, and it will be up to the consume callback function to handle the batch processing.
- maintenance:
     - Cleaned up logging - moved some loggers from debug to silly
     - Fixed some logger's consistency to correctly include "Producer::*" or "Consumer::*" prefixes for clarity
     - Fixed some logger's consistency for events to "Producer::onEvent" or "Consumer::onEvent"
     - Removed commented-out code that is not being used
     - Added unit tests to ensure new features and changes are validated, and also that the test-coverage is met

KNOWN-ISSUES: 
- `Consumer.getMetadata` emits an error `"Consumer::onError()[topics='topic-notification-event'] - Error: Broker: Unknown topic or partition` that does not impact the consumer itself. This seems to be an issue with the underlying Librdkafka library.

### CODE REVIEW

File: src/kafka/consumer.js [hunk 1]
```diff
@@ -259,31 +292,56 @@ class Consumer extends EventEmitter {
       this._consumer = new Kafka.KafkaConsumer(this._config.rdkafkaConf, this._config.topicConf)
 
       this._consumer.setDefaultConsumeTimeout(this._config.options.consumeTimeout)
-      // this._setDefaultConsumeTimeout(this._config.options.consumeTimeout)
+
+      this._consumer.on('warning', warn => {
+        Logger.isDebugEnabled && logger.debug('onWarning', warn)
```
Comment (kleyow) [inline src/kafka/consumer.js:297]: I don't believe logger.level(arg1, arg2) will result in what you want. 
I think you need to jam everything into arg1 which is `message`, but you can double check.

here and throughout.

File: src/kafka/consumer.js [hunk 2]
```diff
@@ -163,17 +163,21 @@ exports.ENUMS = ENUMS
...
+ *     batchSize: 1, // only applicable when mode=1, 2 (i.e. POLL, RECURSIVE)
+ *     pollFrequency: 10, // only applicable for poll mode=1 (i.e. POLL)
+ *     recursiveTimeout: 100, // only applicable when mode=2 (i.e. RECURSIVE)
  *     messageCharset: 'utf8',
  *     messageAsJSON: true,
- *     sync: true
+ *     sync: false,
+ *     syncConcurrency: 1, // only applicable when sync=true
+ *     syncSingleMessage: false, // only applicable when sync=true, and only applicable when mode=2 (i.e. RECURSIVE)
+ *     consumeTimeout: 1000
```
Comment (kleyow) [inline src/kafka/consumer.js:180]:
```suggestion
 *     consumeTimeout: 1000
       deserializeFn: (object, opts) => {}, (optional)
```

File: src/kafka/producer.js [hunk 3]
```diff
@@ -283,31 +352,48 @@ class Producer extends EventEmitter {
...
+        }
+        if (this._producer._isConnecting) {
+          Logger.isDebugEnabled && logger.debug('still connecting')
+        }
+        const parsedMessage = Protocol.parseMessage(messageProtocol)
+        const parsedMessageBuffer = this._config.options.serializeFn(parsedMessage, this._config.options) // this._createBuffer(parsedMessage, this._config.options.messageCharset)
+        if (!parsedMessageBuffer || !(typeof parsedMessageBuffer === 'string' || Buffer.isBuffer(parsedMessageBuffer))) {
+          throw new Error('message must be a string or an instance of Buffer.')
+        }
+        Logger.isDebugEnabled && logger.debug('Producer::send() - start %s', JSON.stringify({
```
Comment (kleyow) [inline src/kafka/producer.js:369]: This type of interpolation won't work due to https://github.com/mojaloop/project/issues/3478
You'll have to use `${}`

File: src/kafka/producer.js [hunk 4]
```diff
@@ -283,31 +352,48 @@ class Producer extends EventEmitter {
...
+          topicName: topicConf.topicName,
+          partition: topicConf.partition,
+          key: topicConf.key
+        }))
+        const producedAt = Date.now()
```
Comment (kleyow) [inline src/kafka/producer.js:374]: nit: Move to line 364 for more ms accuracy?

File: test/perf/run.sh [hunk 5]
```diff
@@ -0,0 +1,41 @@
+#!/bin/bash
```
Comment (kleyow) [inline test/perf/run.sh:1]: Was validating the script on linux. Came accross this. Not sure if this is expected. Attached log file.
(Not really important, unsure why it says unsupported os, the if logic looks sound)

[20230823-15h08m29s.log](https://github.com/mojaloop/central-services-stream/files/12422992/20230823-15h08m29s.log)

```
2023-08-23T20:12:45.038Z - [31merror[39m: Consumer::consume()::syncQueue.queue[0] - workDoneCb - error: Error: KafkaConsumer is disconnected
2023-08-23T20:12:45.038Z - [31merror[39m: Consumer::onError()[topics='test'] - Error: KafkaConsumer is disconnected
    at KafkaConsumer.commitMessageSync (/home/kleyow/mojaloop/central-services-stream/node_modules/node-rdkafka/lib/kafka-consumer.js:581:32)
    at Consumer.commitMessageSync (/home/kleyow/mojaloop/central-services-stream/src/kafka/consumer.js:786:20)
    at /home/kleyow/mojaloop/central-services-stream/test/perf/scripts/consumer.js:135:29
    at new Promise (<anonymous>)
    at /home/kleyow/mojaloop/central-services-stream/test/perf/scripts/consumer.js:111:16
    at /home/kleyow/mojaloop/central-services-stream/src/kafka/consumer.js:473:25
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:4016:13
    at Object.process (/home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1679:21)
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1531:23
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:74:45)
2023-08-23T20:12:45.038Z - [31merror[39m: Consumer::_consumerRecursive()::syncQueue.Single.push - error: Error: KafkaConsumer is disconnected
2023-08-23T20:12:45.038Z - [31merror[39m: Consumer::consume() - onRecursive - error Error: KafkaConsumer is disconnected
2023-08-23T20:12:45.039Z - [31merror[39m: uncaughtException: Local: Erroneous state
Error: Local: Erroneous state
    at Function.createLibrdkafkaError [as create] (/home/kleyow/mojaloop/central-services-stream/node_modules/node-rdkafka/lib/error.js:456:10)
    at KafkaConsumer.Client._errorWrap (/home/kleyow/mojaloop/central-services-stream/node_modules/node-rdkafka/lib/client.js:481:29)
    at KafkaConsumer.resume (/home/kleyow/mojaloop/central-services-stream/node_modules/node-rdkafka/lib/kafka-consumer.js:641:15)
    at /home/kleyow/mojaloop/central-services-stream/src/kafka/consumer.js:484:24
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1496:46
    at Array.forEach (<anonymous>)
    at trigger (/home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1496:27)
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1569:21
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:327:20
    at /home/kleyow/mojaloop/central-services-stream/src/kafka/consumer.js:478:11
2023-08-23T20:12:45.039Z - [31merror[39m: uncaughtException: Local: Erroneous state
Error: Local: Erroneous state
    at Function.createLibrdkafkaError [as create] (/home/kleyow/mojaloop/central-services-stream/node_modules/node-rdkafka/lib/error.js:456:10)
    at KafkaConsumer.Client._errorWrap (/home/kleyow/mojaloop/central-services-stream/node_modules/node-rdkafka/lib/client.js:481:29)
    at KafkaConsumer.resume (/home/kleyow/mojaloop/central-services-stream/node_modules/node-rdkafka/lib/kafka-consumer.js:641:15)
    at /home/kleyow/mojaloop/central-services-stream/src/kafka/consumer.js:484:24
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1496:46
    at Array.forEach (<anonymous>)
    at trigger (/home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1496:27)
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:1569:21
    at /home/kleyow/mojaloop/central-services-stream/node_modules/async/dist/async.js:327:20
    at /home/kleyow/mojaloop/central-services-stream/src/kafka/consumer.js:478:11
```

Comment (mdebarros) [reply test/perf/run.sh:1]: I think we need to jump on a call to discuss this, but I believe that a "state" issue with regard to the offset.

I.e. the offset that your client thinks its at is not the same that its trying to commit.

Comment (mdebarros) [reply test/perf/run.sh:1]: Also, I have not been able to reproduce it on my machine.

Comment (kleyow) [reply test/perf/run.sh:1]: Non-issue. Linux apt installs older version of rdkafka.

**Review by kleyow — APPROVED** [review]

**Review by kleyow — APPROVED** [review]

**Pull request merged**